marker_rustc_driver/src/conversion/marker/expr.rs

```diff
@@ -392,6 +411,54 @@ impl<'ast, 'tcx> MarkerConverterInner<'ast, 'tcx> {
         )
     }
 
+    fn to_expr_from_closure(
+        &self,
+        data: CommonExprData<'ast>,
+        _expr: &hir::Expr<'tcx>,
+        closure: &hir::Closure<'tcx>,
+    ) -> ExprKind<'ast> {
+        let body_id = closure.body;
+        let body = self.rustc_cx.hir().body(body_id);
+        match body.generator_kind {
+            Some(hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Fn)) => {
+                if let hir::ExprKind::Block(block, None) = body.value.kind
+                    && let Some(temp_drop) = block.expr
+                    && let hir::ExprKind::DropTemps(inner_block) = temp_drop.kind
+                {
+                    let api_expr;
+                    super::with_body!(self, body_id, {
+                        api_expr = self.to_expr(inner_block);
+                    });
+                    return api_expr
+                }
+
+                unreachable!("`async fn` body desugar always has the same structure")
+            },
+            Some(hir::GeneratorKind::Async(hir::AsyncGeneratorKind::Block)) => {
+                let block_expr = body.value;
+                if let hir::ExprKind::Block(block, None) = block_expr.kind {
+                    let api_block_expr;
+                    let capture_kind = self.to_capture_kind(closure.capture_clause);
+                    super::with_body!(self, body_id, {
+                        api_block_expr = self.to_block_expr(
+                            CommonExprData::new(self.to_expr_id(block_expr.hir_id), self.to_span_id(block_expr.span)),
+                            block,
+                            None,
+                            Syncness::Async,
+                            capture_kind,
+                        );
+                    });
+                    return ExprKind::Block(self.alloc(api_block_expr));
+                }
+                unreachable!("`async` block desugar always has the same structure")
```

Comment:
I often find `let else` statements less readable. It would be cool if `if-let` chains would be stabilized, but they sadly disable rustftm for the function